src/Microsoft.AspNetCore.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -154,6 +154,117 @@ public static void MergeAttributes(this TagHelperOutput tagHelperOutput, TagBuil
+        /// Adds the given <paramref name="classValue"/> to the <paramref name="tagHelperOutput"/>'s
+        /// <see cref="TagHelperOutput.Attributes"/>.
+        /// </summary>
+        /// <param name="tagHelperOutput">The <see cref="TagHelperOutput"/> this method extends.</param>
+        /// <param name="classValue">The class value to add.</param>
+        public static void AddClass(
+            this TagHelperOutput tagHelperOutput,
+            string classValue)
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            if (!tagHelperOutput.Attributes.TryGetAttribute("class", out TagHelperAttribute classAttribute))
+            {
+                tagHelperOutput.Attributes.Add("class", classValue);
+            }
+            else
+            {
+                var currentClassValue = ExtractClassValue(classAttribute);
+
+                if (currentClassValue.Contains(" ") || currentClassValue.Equals(classValue))
+                {
+                    var arrayOfClasses = currentClassValue.Split(' ');
+                    if (arrayOfClasses.Contains(classValue))
+                    {
+                        return;
+                    }
+                }
+
+                tagHelperOutput.Attributes.SetAttribute(classAttribute.Name,
+                    currentClassValue.Length > 0 ? $"{currentClassValue} {classValue}" : classValue);
+            }
+        }
+
+        /// <summary>
+        /// Removes the given <paramref name="classValue"/> from the <paramref name="tagHelperOutput"/>'s
+        /// <see cref="TagHelperOutput.Attributes"/>.
+        /// </summary>
+        /// <param name="tagHelperOutput">The <see cref="TagHelperOutput"/> this method extends.</param>
+        /// <param name="classValue">The class value to remove.</param>
+        public static void RemoveClass(
+            this TagHelperOutput tagHelperOutput,
+            string classValue)
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            if (!tagHelperOutput.Attributes.TryGetAttribute("class", out TagHelperAttribute classAttribute))
+            {
+                return;
+            }
+
+            var currentClassValue = ExtractClassValue(classAttribute);
+
+            if (string.IsNullOrEmpty(currentClassValue))
+            {
+                return;
+            }
+
+            if (currentClassValue.Contains(" "))
+            {
+                var arrayOfClasses = currentClassValue.Split(' ').ToList();
+
+                if (!arrayOfClasses.Contains(classValue))
+                {
+                    return;
+                }
+
+                arrayOfClasses.RemoveAll(x => x.Equals(classValue));
+                tagHelperOutput.Attributes.SetAttribute(classAttribute.Name, string.Join(" ", arrayOfClasses));
```

Comment:
Again, `currentClassValue` and therefore `arrayOfClasses` can't be used this way. `currentClassValue` is only good for the checks. Even there, it'll fail sometimes.

Preserving what portions need encoding here could get ugly since this also needs to handle `string`, `HtmlString`, `IHtmlContent`, and likely `IHtmlContainer` attribute values differently. Worse, the spaces won't necessarily be appended in the same call as a `classValue` match. Or, the `classValue` could be split between two calls.

Recommend passing the current `HtmlEncoder` in here too. And, set the attribute value to `new HtmlString(string.Join(" ", arrayOfClasses))`.